File: 700.SearchinaBinarySearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
var searchBST = function(root, val) {
    const stack = [root];
    let result = null;
    while(stack.length) {
        const node = stack.pop();
        if (node) {
            if (node.val === val) {
                return node;
            } else if (node.val > val) {
                stack.push(node.left);
            } else {
                stack.push(node.right);
            }
        }
    }
    return result;
};

File: Golang/Tree/700.search-in-a-binary-search-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func searchBSTRecursive(root *TreeNode, val int) *TreeNode {
	if root == nil {
		return nil
	} else if root.Val == val {
		return root
	} else if root.Val > val {
		return searchBSTRecursive(root.Left, val)
	} else {
		return searchBSTRecursive(root.Right, val)
	}
}

func searchBSTPointer(root *TreeNode, val int) *TreeNode {
	node := root
	for node != nil {
		if node.Val == val {
			break
		} else if node.Val > val {
			node = node.Left
		} else {
			node = node.Right
		}
	}
	return node
}
